frontend/test/resolution/testCopyAssignablePrimitives.cpp

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright 2021-2023 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "test-resolution.h"
+
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/resolution/resolution-queries.h"
+#include "chpl/resolution/scope-queries.h"
+#include "chpl/resolution/can-pass.h"
+#include "chpl/types/all-types.h"
+#include "chpl/uast/Identifier.h"
+#include "chpl/uast/Module.h"
+#include "chpl/uast/Record.h"
+#include "chpl/uast/Variable.h"
+
+static constexpr bool testType = true;
+static constexpr bool testExact = !testType;
+
+// args: Vector of argument tuples like (prim, expr, expectedValue,
+// callDotType). callDotType means expr will have '.type' appended.
+// preamble: Code to insert before the generated test code, for declarations.
+static void testPrimitive(
+    std::string preamble,
+    std::vector<std::tuple<const char*, const char*, bool, bool>> args) {
+  Context ctx;
+  auto context = &ctx;
+  /* ErrorGuard guard(context); */
+
+  std::stringstream ps;
+  std::vector<std::string> variables;
+  std::vector<bool> expectedValues;
+
+  ps << preamble << std::endl;
+
+  int counter = 0;
+  for (auto arg : args) {
+    const char* prim = std::get<0>(arg);
+    const char* expr = std::get<1>(arg);
+    bool expectedValue = std::get<2>(arg);
+    bool callDotType = std::get<3>(arg);
+
+    std::string variableName = std::string("x") + std::to_string(counter++);
+    variables.push_back(variableName);
+    expectedValues.push_back(expectedValue);
+
+    // Generate prim call like:
+    // param x3 = __primitive("PRIM_IS_CONST_ASSIGNABLE", (myFoo).type);
+    ps << "param " << variableName << " = " << "__primitive(\"" << prim << "\", (" << expr << ")";
```

Comment:
I'm sure there are some advantages to doing it this way, but IMO it would have been nicer to call `getCopyOrAssignableInfo` directly in this test. Anyway, I don't think it's worth changing now, but if you agree you could leave a TODO comment in this test.

Reply:
I agree for the most part, though a little bit of the implementation of these primitives exists outside of `getCopyOrAssignableInfo` (in `primFamilyCopyableAssignable`, plus primitive resolution in general though we don't need to test it here).

Reply:
I did add the todo though -- as these C++ tests stick around longer-term while more of dyno goes into production, they're probably more valuable as unit tests that complement our large set of regression tests.